Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,197 @@
+import Foundation
+
+/// Represents a MongoDB connection string.
+/// - SeeAlso: https://docs.mongodb.com/manual/reference/connection-string/
+public struct MongoConnectionString: Codable, LosslessStringConvertible {
+    /// Represents a connection string scheme.
+    public struct Scheme: LosslessStringConvertible, Equatable {
+        /// Indicates that this connection string uses the scheme `mongodb`.
+        public static let mongodb = Scheme(.mongodb)
+
+        /// Indicates that this connection string uses the scheme `mongodb+srv`.
+        public static let srv = Scheme(.srv)
+
+        /// Internal representation of a scheme.
+        private enum _Scheme: String {
+            case mongodb
+            case srv = "mongodb+srv"
+        }
+
+        private let _scheme: _Scheme
+
+        private init(_ value: _Scheme) {
+            self._scheme = value
+        }
+
+        /// LosslessStringConvertible` protocol requirements
+        public init?(_ description: String) {
+            guard let _scheme = _Scheme(rawValue: description) else {
+                return nil
+            }
+            self.init(_scheme)
+        }
+
+        public var description: String { self._scheme.rawValue }
+    }
+
+    /// A struct representing a network address, consisting of a host and port.
+    public struct HostIdentifier: Equatable, CustomStringConvertible {
+        private static func verifyPort(from: String, scheme: Scheme) throws -> UInt16? {
```

Comment:
I think it's probably simpler if we constrain the functionality of this method to just parsing the port from a string (and we can call it something to reflect that e.g. `parsePort`). As it is now, it's hard to infer from the name of the method and the arguments exactly what this method does.

Similarly, I think we can simplify things by not passing the scheme to the `HostIdentifier` initializer that parses the string, and instead have that initializer only focus on parsing strings. We can then enforce the SRV scheme constraints (i.e. only a single host with no port) in the `MongoConnectionString` initializer, since that's its area of concern.

Reply:
done